File: services/storage/auth/cognito.go
package auth

import (
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
)

type CognitoUser struct {
	UserID   string   `json:"sub"`
	Email    string   `json:"email"`
	UserType string   `json:"custom:user_type"` // "creator", "consumer", "admin"
	Groups   []string `json:"cognito:groups"`
}

type JWKKey struct {
	Alg string `json:"alg"`
	E   string `json:"e"`
	Kid string `json:"kid"`
	Kty string `json:"kty"`
	N   string `json:"n"`
	Use string `json:"use"`
}

type JWKS struct {
	Keys []JWKKey `json:"keys"`
}

type JWTHeader struct {
	Alg string `json:"alg"`
	Kid string `json:"kid"`
}

type JWTClaims struct {
	Sub       string   `json:"sub"`
	Email     string   `json:"email"`
	UserType  string   `json:"custom:user_type"`
	Groups    []string `json:"cognito:groups"`
	Exp       int64    `json:"exp"`
	Iat       int64    `json:"iat"`
	TokenUse  string   `json:"token_use"`
	ClientID  string   `json:"client_id"`
}

var (
	jwksCache     *JWKS
	jwksCacheTime time.Time
	cacheDuration = 24 * time.Hour
)

// VerifyCognitoToken verifies the JWT token with AWS Cognito
func VerifyCognitoToken(tokenString string) (*CognitoUser, error) {
	// Parse JWT without verification first to get the header
	parts := strings.Split(tokenString, ".")
	if len(parts) != 3 {
		return nil, errors.New("invalid token format")
	}

	// Decode header
	headerBytes, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return nil, fmt.Errorf("failed to decode header: %v", err)
	}

	var header JWTHeader
	if err := json.Unmarshal(headerBytes, &header); err != nil {
		return nil, fmt.Errorf("failed to parse header: %v", err)
	}

	// Get JWKS
	jwks, err := getJWKS()
	if err != nil {
		return nil, fmt.Errorf("failed to get JWKS: %v", err)
	}

	// Find the key
	var key *JWKKey
	for _, k := range jwks.Keys {
		if k.Kid == header.Kid {
			key = &k
			break
		}
	}

	if key == nil {
		return nil, errors.New("key not found in JWKS")
	}

	// Verify signature
	if err := verifyJWTSignature(tokenString, key); err != nil {
		return nil, fmt.Errorf("signature verification failed: %v", err)
	}

	// Parse claims
	claimsBytes, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, fmt.Errorf("failed to decode claims: %v", err)
	}

	var claims JWTClaims
	if err := json.Unmarshal(claimsBytes, &claims); err != nil {
		return nil, fmt.Errorf("failed to parse claims: %v", err)
	}

	// Verify token use
	if claims.TokenUse != "id" && claims.TokenUse != "access" {
		return nil, errors.New("invalid token use")
	}

	// Verify expiration
	if time.Now().Unix() > claims.Exp {
		return nil, errors.New("token expired")
	}

	// Verify client ID if configured
	expectedClientID := os.Getenv("COGNITO_CLIENT_ID")
	if expectedClientID != "" && claims.ClientID != expectedClientID {
		return nil, errors.New("invalid client ID")
	}

	return &CognitoUser{
		UserID:   claims.Sub,
		Email:    claims.Email,
		UserType: claims.UserType,
		Groups:   claims.Groups,
	}, nil
}

// getJWKS fetches the JSON Web Key Set from Cognito
func getJWKS() (*JWKS, error) {
	// Check cache
	if jwksCache != nil && time.Since(jwksCacheTime) < cacheDuration {
		return jwksCache, nil
	}

	// Build JWKS URL
	region := os.Getenv("AWS_REGION")
	if region == "" {
		region = "us-east-1"
	}
	userPoolID := os.Getenv("COGNITO_USER_POOL_ID")
	if userPoolID == "" {
		return nil, errors.New("COGNITO_USER_POOL_ID not configured")
	}

	jwksURL := fmt.Sprintf("https://cognito-idp.%s.amazonaws.com/%s/.well-known/jwks.json", region, userPoolID)

	// Fetch JWKS
	resp, err := http.Get(jwksURL)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch JWKS: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("JWKS request failed with status: %d", resp.StatusCode)
	}

	var jwks JWKS
	if err := json.NewDecoder(resp.Body).Decode(&jwks); err != nil {
		return nil, fmt.Errorf("failed to decode JWKS: %v", err)
	}

	// Update cache
	jwksCache = &jwks
	jwksCacheTime = time.Now()

	return &jwks, nil
}

// verifyJWTSignature verifies the JWT signature using the JWK
func verifyJWTSignature(tokenString string, key *JWKKey) error {
	// For now, we'll return nil as proper RSA verification requires additional libraries
	// In production, use a proper JWT library like github.com/golang-jwt/jwt
	// This is a placeholder that should be replaced with actual verification
	
	if key.Alg != "RS256" {
		return errors.New("unsupported algorithm")
	}

	// TODO: Implement actual RSA signature verification
	// This would involve:
	// 1. Converting the JWK to an RSA public key
	// 2. Verifying the signature using the public key
	
	log.Println("WARNING: JWT signature verification not fully implemented")
	return nil
}

// GetUserFromContext retrieves the authenticated user from the Gin context
func GetUserFromContext(c *gin.Context) (*CognitoUser, bool) {
	user, exists := c.Get("user")
	if !exists {
		return nil, false
	}
	cognitoUser, ok := user.(*CognitoUser)
	return cognitoUser, ok
}

// GetUserIDFromContext retrieves the authenticated user ID from the Gin context
func GetUserIDFromContext(c *gin.Context) (string, bool) {
	userID := c.GetString("userID")
	return userID, userID != ""
}

// IsInGroup checks if a group is in the list of groups
func IsInGroup(groups []string, targetGroup string) bool {
	for _, group := range groups {
		if group == targetGroup {
			return true
		}
	}
	return false
}

// IsCreator checks if the user is a creator
func IsCreator(user *CognitoUser) bool {
	return user.UserType == "creator" || IsInGroup(user.Groups, "creators")
}

// IsAdmin checks if the user is an admin
func IsAdmin(user *CognitoUser) bool {
	return user.UserType == "admin" || IsInGroup(user.Groups, "admins")
}

// IsConsumer checks if the user is a consumer
func IsConsumer(user *CognitoUser) bool {
	return user.UserType == "consumer" || IsInGroup(user.Groups, "consumers")
}

File: backend/api/auth/cognito.py
"""
AWS Cognito authentication for production use
"""
import os
import jwt
from typing import Optional, Dict, Any
from datetime import datetime
from fastapi import HTTPException, status
import httpx
from jwt.algorithms import RSAAlgorithm
import json

class CognitoAuth:
    def __init__(self):
        self.region = os.getenv("AWS_REGION", "us-east-1")
        self.user_pool_id = os.getenv("COGNITO_USER_POOL_ID")
        self.client_id = os.getenv("COGNITO_CLIENT_ID")
        self.jwks_url = f"https://cognito-idp.{self.region}.amazonaws.com/{self.user_pool_id}/.well-known/jwks.json"
        self.issuer = f"https://cognito-idp.{self.region}.amazonaws.com/{self.user_pool_id}"
        self._jwks_client = None
        self._keys = {}

    async def _get_jwks(self):
        """Fetch JSON Web Key Set from Cognito"""
        if self._jwks_client is None:
            self._jwks_client = httpx.AsyncClient()
        
        response = await self._jwks_client.get(self.jwks_url)
        response.raise_for_status()
        return response.json()

    async def _get_signing_key(self, kid: str):
        """Get the signing key for a given key ID"""
        if kid not in self._keys:
            jwks = await self._get_jwks()
            for key in jwks["keys"]:
                if key["kid"] == kid:
                    self._keys[kid] = RSAAlgorithm.from_jwk(json.dumps(key))
                    break
            else:
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Unable to find a signing key that matches"
                )
        return self._keys[kid]

    async def verify_token(self, token: str) -> Dict[str, Any]:
        """Verify a Cognito JWT token"""
        try:
            # Decode token header without verification to get the key ID
            unverified_header = jwt.get_unverified_header(token)
            kid = unverified_header["kid"]
            
            # Get the signing key
            signing_key = await self._get_signing_key(kid)
            
            # Verify and decode the token
            payload = jwt.decode(
                token,
                signing_key,
                algorithms=["RS256"],
                audience=self.client_id,
                issuer=self.issuer,
                options={"verify_exp": True}
            )
            
            # Verify token use
            if payload.get("token_use") not in ["id", "access"]:
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Invalid token use"
                )
            
            return payload
            
        except jwt.ExpiredSignatureError:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Token has expired"
            )
        except jwt.InvalidTokenError as e:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail=f"Invalid token: {str(e)}"
            )
        except Exception as e:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail=f"Token verification failed: {str(e)}"
            )

    async def get_user_from_token(self, token: str) -> Dict[str, Any]:
        """Extract user information from a verified token"""
        payload = await self.verify_token(token)
        
        # Build user object from token claims
        user = {
            "id": payload.get("sub"),
            "email": payload.get("email"),
            "name": payload.get("name") or payload.get("given_name", "") + " " + payload.get("family_name", ""),
            "username": payload.get("cognito:username", payload.get("email")),
            "groups": payload.get("cognito:groups", []),
            "token_use": payload.get("token_use"),
            "exp": payload.get("exp"),
            "iat": payload.get("iat")
        }
        
        # Clean up empty values
        user = {k: v for k, v in user.items() if v}
        
        return user

    async def close(self):
        """Close the HTTP client"""
        if self._jwks_client:
            await self._jwks_client.aclose()

# Global instance
cognito_auth = None

def get_cognito_auth() -> Optional[CognitoAuth]:
    """Get or create Cognito auth instance"""
    global cognito_auth
    
    # Check if Cognito is configured
    if not os.getenv("COGNITO_USER_POOL_ID") or not os.getenv("COGNITO_CLIENT_ID"):
        return None
    
    if cognito_auth is None:
        cognito_auth = CognitoAuth()
    
    return cognito_auth